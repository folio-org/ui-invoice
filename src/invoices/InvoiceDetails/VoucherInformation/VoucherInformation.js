import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';

import { stripesConnect } from '@folio/stripes/core';
import {
  Col,
  Headline,
  KeyValue,
  MultiColumnList,
  Row,
} from '@folio/stripes/components';
import { AmountWithCurrencyField } from '@folio/stripes-acq-components';

import { formatDate } from '../../../common/utils';

const visibleColumns = ['externalAccountNumber', 'amount'];
const columnMapping = {
  externalAccountNumber: <FormattedMessage id="ui-invoice.invoice.details.voucher.externalAccountNumber" />,
  amount: <FormattedMessage id="ui-invoice.invoice.details.voucher.amount" />,
};
const columnWidths = {
  externalAccountNumber: '50%',
  amount: '50%',
};

const VoucherInformation = ({ voucher, voucherLines = [] }) => {
  const resultsFormatter = {
    amount: line => (
      <AmountWithCurrencyField
        amount={line.amount}
        currency={voucher.invoiceCurrency}
      />
    ),
  };

  return (
    <Fragment>
      <Row>
        <Col xs={3}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoice.details.voucher.status" />}
            value={voucher.status}
          />
        </Col>

        <Col xs={3}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoice.details.voucher.voucherNumber" />}
            value={voucher.voucherNumber}
          />
        </Col>

        <Col xs={3}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoice.details.voucher.voucherDate" />}
            value={formatDate(voucher.voucherDate)}
          />
        </Col>

        <Col xs={3}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoice.details.voucher.total" />}
          >
            <AmountWithCurrencyField
              amount={voucher.amount}
              currency={voucher.invoiceCurrency}
            />
          </KeyValue>
        </Col>
      </Row>

      <Row>
        <Col xs={3}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoice.details.voucher.exchangeRate" />}
            value={voucher.exchangeRate}
          />
        </Col>

        <Col xs={3}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoice.details.voucher.disbursementNumber" />}
            value={voucher.disbursementNumber}
          />
        </Col>

        <Col xs={3}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoice.details.voucher.disbursementDate" />}
            value={formatDate(voucher.disbursementDate)}
          />
        </Col>

        <Col xs={3}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoice.details.voucher.disbursementAmount" />}
          >
            <AmountWithCurrencyField
              amount={voucher.disbursementAmount}
              currency={voucher.invoiceCurrency}
            />
          </KeyValue>
        </Col>
      </Row>

      <Row>
        <Col xs={3}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoice.details.voucher.accountingCode" />}
            value={voucher.accountingCode}
          />
        </Col>
      </Row>

      <Headline
        margin="none"
        size="large"
      >
        <FormattedMessage id="ui-invoice.invoice.details.voucher.voucherLines" />
      </Headline>

      <MultiColumnList
        columnMapping={columnMapping}
        columnWidths={columnWidths}
        contentData={voucherLines}
        formatter={resultsFormatter}
        visibleColumns={visibleColumns}
      />
    </Fragment>
  );
};

VoucherInformation.propTypes = {
  voucher: PropTypes.object.isRequired,
  voucherLines: PropTypes.arrayOf(PropTypes.object),
};

export default stripesConnect(VoucherInformation);
