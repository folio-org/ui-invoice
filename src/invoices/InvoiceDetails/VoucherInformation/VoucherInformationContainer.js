import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { get } from 'lodash';

import { stripesConnect } from '@folio/stripes/core';
import { Icon } from '@folio/stripes/components';

import {
  VOUCHER,
  VOUCHER_LINES,
} from '../../../common/resources';
import VoucherInformation from './VoucherInformation';

const VoucherInformationContainer = ({ invoiceId, mutator, resources }) => {
  const voucher = get(resources, ['voucher', 'records', 0], {});
  const voucherLines = get(resources, ['voucherLines', 'records', 0], []);
  const isLoading = !get(resources, ['voucher', 'hasLoaded']);

  useEffect(() => {
    mutator.voucher.reset();
    // mutator.voucherLines.reset();

    mutator.voucher.GET().then(response => {
      const voucherId = response[0].id;

      if (voucherId) {
        mutator.voucherLines.GET({
          params: {
            query: `voucherId==${voucherId}`,
          },
        });
      }
    });
  }, [invoiceId]);

  if (isLoading) {
    return (
      <Icon
        icon="spinner-ellipsis"
        width="100px"
      />
    );
  }

  return (
    <VoucherInformation
      voucher={voucher}
      voucherLines={voucherLines}
    />
  );
};

VoucherInformationContainer.manifest = {
  voucher: VOUCHER,
  voucherLines: {
    ...VOUCHER_LINES,
    fetch: false,
    accumulate: true,
  },
};

VoucherInformationContainer.propTypes = {
  // invoiceId prop is used in manifest
  // eslint-disable-next-line react/no-unused-prop-types
  invoiceId: PropTypes.string.isRequired,
  mutator: PropTypes.shape({
    voucher: PropTypes.object.isRequired,
    voucherLines: PropTypes.object.isRequired,
  }).isRequired,
  resources: PropTypes.object.isRequired,
};

export default stripesConnect(VoucherInformationContainer);
