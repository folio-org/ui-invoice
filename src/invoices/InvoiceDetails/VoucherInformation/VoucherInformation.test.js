import React from 'react';
import { render } from '@testing-library/react';

import '../../../../test/jest/__mock__';

import { batchVoucher, batchVoucherLine } from '../../../../test/jest/fixtures';

import VoucherInformation from './VoucherInformation';

jest.mock('../../Voucher/VoucherDetails', () => jest.fn().mockReturnValue('VoucherDetails'));

const defaultProps = {
  voucher: batchVoucher,
  voucherLines: [batchVoucherLine],
};
const renderVoucherInformation = (props = defaultProps) => render(
  <VoucherInformation {...props} />,
);

describe('VoucherInformation', () => {
  beforeEach(() => {
    global.document.createRange = global.document.originalCreateRange;
  });

  afterEach(() => {
    global.document.createRange = global.document.mockCreateRange;
  });

  it('should render correct structure for invoice voucher info', () => {
    const { asFragment } = renderVoucherInformation();

    expect(asFragment()).toMatchSnapshot();
  });
});
