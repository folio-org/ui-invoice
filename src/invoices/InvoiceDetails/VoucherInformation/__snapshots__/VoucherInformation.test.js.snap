// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VoucherInformation should render correct structure for invoice voucher info 1`] = `
<DocumentFragment>
  VoucherDetails
  <div
    class="headline size-large margin-none font-weight-bold"
    data-test-headline="true"
  >
    ui-invoice.invoice.details.voucher.voucherLines
  </div>
  <div
    aria-live="assertive"
    aria-relevant="additions"
    class="sr-only"
    tabindex="-1"
  />
  <div
    aria-rowcount="2"
    class="mclContainer"
    data-total-count="0"
    role="grid"
    style="position: relative; overflow: hidden; height: auto; width: 100%;"
    tabindex="0"
  >
    <div
      class="mclHeaderContainer"
    >
      <div
        aria-rowindex="1"
        class="mclHeaderRow"
        role="row"
        style="display: flex;"
      >
        <div
          aria-sort="none"
          class="mclHeader"
          id="list-column-externalaccountnumber"
          role="columnheader"
          style="flex: 0 0 auto;"
        >
          ui-invoice.invoice.details.voucher.externalAccountNumber
        </div>
        <div
          aria-sort="none"
          class="mclHeader"
          id="list-column-amount"
          role="columnheader"
          style="flex: 0 0 auto;"
        >
          ui-invoice.invoice.details.voucher.amount
        </div>
      </div>
    </div>
    <div
      class="mclScrollable"
      style="position: static; width: 100%;"
      tabindex="0"
    >
      <div
        class="mclRowContainer"
        role="rowgroup"
        style="position: static;"
      >
        <div
          aria-rowindex="2"
          class="mclRowFormatterContainer"
          data-row-index="row-0"
          role="row"
          style="position: static;"
        >
          <div
            class="mclRow mclIsOdd mclIsInteractive"
            data-row-inner="0"
            style="min-width: 0px;"
          >
            <div
              class="mclCell  mclCellStyle"
              role="gridcell"
            >
              54321098
            </div>
            <div
              class="mclCell  mclCellStyle"
              role="gridcell"
            >
              $23.45
            </div>
          </div>
        </div>
        <div
          class="mclEndOfListContainer"
          data-end-of-list="1"
        >
          <div
            class="mclCenteredContainer"
            style="width: 100%; visibility: hidden; height: 0px; padding: 0px;"
          >
            <div
              class="textCentered"
              data-test-layout="true"
            >
              <div
                class="mclEndOfList"
                data-test-layout="true"
              >
                <span>
                  Icon
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
